server/src/main/java/org/elasticsearch/common/Rounding.java

```diff
@@ -1132,50 +1132,69 @@ public long beforeOverlap(long localMillis, Overlap overlap) {
          */
         private class JavaTimeRounding extends TimeIntervalPreparedRounding {
             @Override
-            public long round(long utcMillis) {
-                final Instant utcInstant = Instant.ofEpochMilli(utcMillis);
-                final LocalDateTime rawLocalDateTime = LocalDateTime.ofInstant(utcInstant, timeZone);
-
-                // a millisecond value with the same local time, in UTC, as `utcMillis` has in `timeZone`
-                final long localMillis = utcMillis + timeZone.getRules().getOffset(utcInstant).getTotalSeconds() * 1000;
-                assert localMillis == rawLocalDateTime.toInstant(ZoneOffset.UTC).toEpochMilli();
-
-                final long roundedMillis = roundKey(localMillis, interval) * interval;
-                final LocalDateTime roundedLocalDateTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(roundedMillis), ZoneOffset.UTC);
-
-                // Now work out what roundedLocalDateTime actually means
-                final List<ZoneOffset> currentOffsets = timeZone.getRules().getValidOffsets(roundedLocalDateTime);
-                if (currentOffsets.isEmpty() == false) {
-                    // There is at least one instant with the desired local time. In general the desired result is
-                    // the latest rounded time that's no later than the input time, but this could involve rounding across
-                    // a timezone transition, which may yield the wrong result
-                    final ZoneOffsetTransition previousTransition = timeZone.getRules().previousTransition(utcInstant.plusMillis(1));
-                    for (int offsetIndex = currentOffsets.size() - 1; 0 <= offsetIndex; offsetIndex--) {
-                        final OffsetDateTime offsetTime = roundedLocalDateTime.atOffset(currentOffsets.get(offsetIndex));
-                        final Instant offsetInstant = offsetTime.toInstant();
-                        if (previousTransition != null && offsetInstant.isBefore(previousTransition.getInstant())) {
-                            /*
-                             * Rounding down across the transition can yield the
-                             * wrong result. It's best to return to the transition
-                             * time and round that down.
-                             */
-                            return round(previousTransition.getInstant().toEpochMilli() - 1);
+            public long round(long originalUtcMillis) {
+                long utcMillis = originalUtcMillis;
+                int attempts = 0;
+                /* 
+                 * We give up after 5000 attempts and throw an exception. The
+                 * most attempts I could get running locally are 500 - for
+                 * Asia/Tehran with an 80,000 day range. You just can't declare
+                 * ranges much larger than that in ES right now.   
+                 */
+                attempt: while (attempts < 5000) {
```

Comment:
I wonder if we can make this constant a field that can be modified with the builder which defaults to 5000? That way we can write a unit test that ensures we correctly throw the exception in case the loop limit is reached.

Reply:
:+1: